packages/framework/src/client/__tests__/resource-request.spec.js

```diff
@@ -0,0 +1,218 @@
+			const promise = new Promise( ( resolve ) => {
+				promiseResolve = resolve;
+			} );
+
+			request.requested( promise );
+			expect( request.getStatus() ).toBe( STATUS.inFlight );
+
+			promiseResolve();
+
+			return promise.then( () => {
+				expect( request.getStatus() ).toBe( STATUS.complete );
+			} );
+		} );
+
+		it( 'progresses through error flow correctly', () => {
+			const request = new ResourceRequest( 'resource1', {}, {}, tenSecondsAgo );
+
+			expect( request.getStatus( now ) ).toBe( STATUS.overdue );
+
+			let promiseReject;
+			const promise = new Promise( ( resolve, reject ) => {
+				promiseReject = reject;
+			} );
+
+			request.requested( promise, now );
+			expect( request.getStatus( now ) ).toBe( STATUS.inFlight );
+
+			promiseReject( 'error message' );
+
+			return promise.catch( () => {
+				expect( request.getStatus( now ) ).toBe( STATUS.failed );
+				expect( request.error ).toBe( 'error message' );
+			} );
+		} );
+	} );
+
+	describe( 'addRequirement', () => {
+		it( 'reschedules earlier', () => {
+			const requirement1 = { freshness: 5 * MINUTE, timeout: 3 * SECOND };
+			const requirement2 = { freshness: 4 * MINUTE, timeout: 3 * SECOND };
+			const resourceState = { lastReceived: threeMinutesAgo };
+
+			const request = new ResourceRequest( 'resource1', requirement1, resourceState, tenSecondsAgo );
+			expect( request.getStatus( now ) ).toBe( STATUS.scheduled );
+			expect( request.getTimeLeft( now ) ).toBe( 2 * MINUTE );
+
+			expect( request.addRequirement( requirement2, resourceState, now ) ).toBeTruthy();
+			expect( request.getStatus( now ) ).toBe( STATUS.scheduled );
+			expect( request.getTimeLeft( now ) ).toBe( 1 * MINUTE );
+		} );
+
+		it( 'does not reschedule if request is already adequate', () => {
+			const requirement1 = { freshness: 4 * MINUTE, timeout: 3 * SECOND };
+			const requirement2 = { freshness: 5 * MINUTE, timeout: 3 * SECOND };
+			const resourceState = { lastReceived: threeMinutesAgo };
+
+			const request = new ResourceRequest( 'resource1', requirement1, resourceState, tenSecondsAgo );
+			expect( request.getStatus( now ) ).toBe( STATUS.scheduled );
+			expect( request.getTimeLeft( now ) ).toBe( 1 * MINUTE );
+
+			expect( request.addRequirement( requirement2, resourceState, now ) ).toBeTruthy();
+			expect( request.getStatus( now ) ).toBe( STATUS.scheduled );
+			expect( request.getTimeLeft( now ) ).toBe( 1 * MINUTE );
+		} );
+
+		it( 'fails if the request is not scheduled', () => {
+			const requirement1 = { timeout: 3 * SECOND };
+			const requirement2 = { freshness: 4 * MINUTE, timeout: 3 * SECOND };
+			const resourceState = { lastReceived: threeMinutesAgo };
+
+			const request = new ResourceRequest( 'resource1', requirement1, resourceState, tenSecondsAgo );
+			expect( request.getStatus( now ) ).toBe( STATUS.unnecessary );
+
+			expect( request.addRequirement( requirement2, resourceState, now ) ).toBeFalsy();
+			expect( request.getStatus( now ) ).toBe( STATUS.unnecessary );
```

Comment:
Is this equivalent to two different components needing the same resource? If so, wouldn't we want to follow the freshness of the second requirement even if the first only needed the resource once?

Reply:
In this case, `requirement1` has no freshness requirement, so it's happy to have whatever data exists, as long as it's been fetched at least once in the past. `requirement2` wants the data to exist and be received no longer than 4 minutes ago. Both of those requirements are already fulfilled by the current `resourceState` so the request remains unnecessary.